compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -169,6 +156,594 @@ bool TR_CFGSimplifier::simplifyBooleanStore()
    if (!(_next2->getPredecessors().front()->getFrom() == _block && (_next2->getPredecessors().size() == 1)))
       needToDuplicateTree = true;
 
+   // This block must end in a compare-and-branch which can be converted to a
+   // boolean compare, or a branch using the condition code.
+   //
+   TR::TreeTop *compareTreeTop = getLastRealTreetop(_block);
+   TR::Node *compareNode       = compareTreeTop->getNode();
+   if (!compareNode->getOpCode().isIf())
+      return false;
+   if (compareNode->isNopableInlineGuard())
+      //don't simplify nopable guards
+      return false;
+
+   // ... and so one of the successors must be the fall-through successor. Make
+   // _next1 be the fall-through successor.
+   //
+   TR::Block *b = getFallThroughBlock(_block);
+   if (b != _next1)
+      {
+      TR_ASSERT(b == _next2, "CFG Simplifier");
+      _next2 = _next1;
+      _next1 = b;
+      }
+   return simplifyIfPatterns(needToDuplicateTree);
+   }
+
+bool OMR::CFGSimplifier::simplifyIfPatterns(bool needToDuplicateTree)
+   {
+   return simplifyBooleanStore(needToDuplicateTree)
+          || simplifyNullToException(needToDuplicateTree)
+          || simplifySimpleStore(needToDuplicateTree)
+          || simplifyCondStoreSequence(needToDuplicateTree)
+          || simplifyInstanceOfTestToCheckcast(needToDuplicateTree)
+          ;   
+   }
+
+bool  OMR::CFGSimplifier::hasExceptionPoint(TR::Block *block, TR::TreeTop *end)
+   {
+   return true;
+   if (!block->getExceptionSuccessors().empty())
+      return true;
+
+   for (TR::TreeTop *itr = block->getEntry(); itr && itr != end; itr = itr->getNextTreeTop())
+       {
+       if (itr->getNode()->exceptionsRaised() != 0)
+          return true;
+       }
+   return false;
+   }
+
+bool OMR::CFGSimplifier::simplifyInstanceOfTestToCheckcast(bool needToDuplicateTree)
+   {
+   static char *disableSimplifyInstanceOfTestToCheckcast = feGetEnv("TR_disableSimplifyInstanceOfTestToCheckcast");
```

Comment:
Is there a reason for using an environment variable instead of an option here?

Reply:
These env variables are used to disable matching individual patterns within the CFG simplification opt. There is a jit option `disableCFGSimplification` that disable the entire opt. I think it's appropriate to use env vars to selectively disable specific patterns for bebugging and perf measurements; it seems a bit too fine grained to have to expose sub-optimization internals, e.g. individual CFG patterns, in full fledged jit options.

Reply:
I would argue that debugging and measuring perf are all valid use cases for compiler options. From that perspective, I'm not sure I see much difference between exposing the ability to disable optimization internals via environment variables vs compiler options, other than options are generally cleaner and easier to keep track of.

That said, I see lot's of other places in the optimizer that use `feGetEnv()`. Are environment variables the conventional way of controlling optimization internals?

Reply:
There are many places where env variables are used to control optmizer/codegen behavior, some are used to back off to a different, sometimes conservative, path, others are used to disable, usually small, part of an opt. Looking at the options file, it seems to me we tend to use option to toggle entire opt or a significant feature of an opt (like disableInlinng, or disableTraps, disableFanin), and leave smaller control to env variables. To me the env variables here are controlling smaller sub-optimization logic, so I felt they don't warrant full options.

Reply:
There is definitely precedent for these finer gained controls to be environment variables. Arguably which patterns to turn on and off and how they interact with each other is an internal opt detail. exposing all of these things as full options will bloat the option set. I believe this fits with current optimizer style and is an appropriate balance between exposing controls and providing debug/tuning handles.

Reply:
Alright. If there is a good president for using environment variables this way, then I think it makes sense to use them here as well.